aten/src/ATen/Error.h

```diff
@@ -5,14 +5,164 @@
+  /// whatever (possibly mangled) name we got from `backtrace()`.
+  std::string function_name;
+  /// This is a number in hexadecimal form (e.g. "0xdead") representing the
+  /// offset into the function's machine code at which the function's body
+  /// starts, i.e. skipping the "prologue" that handles stack manipulation and
+  /// other calling convention things.
+  std::string offset_into_function;
+  /// NOTE: In debugger parlance, the "object file" refers to the ELF file that
+  /// the symbol originates from, i.e. either an executable or a library.
+  std::string object_file;
+};
+
+inline FrameInformation parse_frame_information(
+    const std::string& frame_string) {
+  FrameInformation frame;
+
+  // This is the function name in the CXX ABI mangled format, e.g. something
+  // like _Z1gv. Reference:
+  // https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling
+  std::string mangled_function_name;
+
+#ifdef __GLIBCXX__
+  // In GLIBCXX, `frame_string` follows the pattern
+  // `<object-file>(<mangled-function-name>+<offset-into-function>)
+  // [<return-address>]`
+
+  const auto function_name_start = frame_string.find('(') + 1;
+  const auto offset_start = frame_string.find('+', function_name_start) + 1;
+  const auto offset_end = frame_string.find(')', offset_start);
+  frame.object_file = frame_string.substr(0, function_name_start - 1);
+  frame.offset_into_function =
+      frame_string.substr(offset_start, offset_end - offset_start);
+
+  // NOTE: We don't need to parse the return address because
+  // we already have it from the call to `backtrace()`.
+
+  mangled_function_name = frame_string.substr(
+      function_name_start, (offset_start - 1) - function_name_start);
+#else
+  // In LIBCXX, The pattern is
+  // `<frame number> <object-file> <return-address> <mangled-function-name> +
+  // <offset-into-function>`
+  std::string skip;
+  std::istringstream input_stream(frame_string);
+  input_stream >> skip >> frame.object_file >> skip >> mangled_function_name >>
+      skip >> frame.offset_into_function;
+#endif
+
+  int status = -1;
+  // This function will demangle the mangled function name into a more human
+  // readable format, e.g. _Z1gv -> g().
+  // More information:
+  // https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/libsupc%2B%2B/cxxabi.h
+  char* demangled_function_name = abi::__cxa_demangle(
+      mangled_function_name.c_str(),
+      /*__output_buffer=*/nullptr,
+      /*__length=*/0,
+      &status);
+
+  // Demangling may fail, for example when the name does not follow the
+  // standard C++ (Itanium ABI) mangling scheme. This is the case for `main` or
+  // `clone` for example, so the mangled name is a fine default.
+  if (status == 0) {
+    frame.function_name = demangled_function_name;
+  } else {
+    frame.function_name = mangled_function_name;
+  }
+
+  // `__cxa_demangle` returns a malloc'd string that we have to free ourselves.
+  free(demangled_function_name);
```

Comment:
Can you use `unique_ptr` for this? There are allocations that can fail between those lines

Reply:
Yeah good point, should've thought of this